src/main/java/com/microsoft/azure/relay/TokenRenewer.java

```diff
@@ -5,13 +5,17 @@
 import java.util.Timer;
 import java.util.TimerTask;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.Consumer;
 
 class TokenRenewer {
 	private final HybridConnectionListener listener;
 	private final String appliesTo;
 	private final Duration tokenValidFor;
+	private final Lock scheduleLock = new ReentrantLock();
```

Comment:
>ReentrantLock [](start = 39, length = 13)

Is there a reason to use ReentrantLock instead of synchronized(this) or whatever?
``` JAVA
public class DemoClass
{
    public synchronized void demoMethod(){}
}
 
or
 
public class DemoClass
{
    public void demoMethod(){
        synchronized (this)
        {
            //other thread safe code
        }
    }
}
```
[https://howtodoinjava.com/java/multi-threading/object-vs-class-level-locking/](https://howtodoinjava.com/java/multi-threading/object-vs-class-level-locking/)

Reply:
From [https://howtodoinjava.com/java/multi-threading/how-to-use-locks-in-java-java-util-concurrent-locks-lock-tutorial-and-example/](https://howtodoinjava.com/java/multi-threading/how-to-use-locks-in-java-java-util-concurrent-locks-lock-tutorial-and-example/):

Difference between Lock Interface and synchronized keyword
The main differences between a Lock and a synchronized block are:

1) Having a timeout trying to get access to a synchronized block is not possible. Using Lock.tryLock(long timeout, TimeUnit timeUnit), it is possible.
2) The synchronized block must be fully contained within a single method. A Lock can have it’s calls to lock() and unlock() in separate methods.

---
In reply to: [338833472](https://github.com/Azure/azure-relay-java/pull/54#discussion_r338833472) [](ancestors = 338833472)

Reply:
I mostly wanted ```.tryLock()```, which just skips the block instead of ```synchronized()```, which sits there and block while wait for the lock

---
In reply to: [338833472](https://github.com/Azure/azure-relay-java/pull/54#discussion_r338833472) [](ancestors = 338833472)

Reply:
Ah, I see.

---
In reply to: [338836278](https://github.com/Azure/azure-relay-java/pull/54#discussion_r338836278) [](ancestors = 338836278,338833472)